### PR TITLE
gh-108014: Add Py_IsFinalizing() function

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-108014

----
:books: Documentation preview :books:: https://cpython-previews--108032.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
I prepared https://github.com/python/pythoncapi-compat/pull/66 to add the function to the pythoncapi-compat project.

File: Doc/c-api/init.rst [hunk 1]
```diff
@@ -373,6 +373,14 @@ Initializing and finalizing the interpreter
    :c:func:`Py_Initialize` is called again.
 
 
+.. c:function:: int Py_IsFinalizing()
+
+   Return non-zero if the Python interpreter is :term:`shutting down
+   <interpreter shutdown>`, return 0 otherwise.
```
Comment (ericsnowcurrently) [inline Doc/c-api/init.rst]: Is this referring to the runtime (AKA main interpreter) or to the current interpreter?  I'd expect it to be the current interpreter.

Would it make sense to explicitly pass an interpreter and use `NULL` to mean the current one (or main)?

Comment (vstinner) [reply Doc/c-api/init.rst]: I copied the documentation of the sys.is_finalizing() function. If you want a different feature (pass an interpreter), I would prefer to have a different function.

It's set by Py_Finalize() which finalizes the main interpreter: other sub-interpreters should be deleted before explicitly, no?

File: Doc/c-api/init.rst [hunk 2]
```diff
@@ -373,6 +373,14 @@ Initializing and finalizing the interpreter
    :c:func:`Py_Initialize` is called again.
 
 
+.. c:function:: int Py_IsFinalizing()
```
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst:376]: What is the order of functions here, lexicographical or functional? Should not it be below `Py_FinalizeEx()`?

File: Doc/c-api/init.rst (hunk 1 above)
Comment (serhiy-storchaka) [inline Doc/c-api/init.rst]: For consistency with `Py_IsInitialized()`:
```suggestion
   Return true (nonzero) if the Python interpreter is :term:`shutting down
   <interpreter shutdown>`, false (zero) if not.
```

File: Doc/c-api/init.rst (hunk 2 above)
Comment (vstinner) [reply Doc/c-api/init.rst:376]: For me, the order is more logical:

* Functions to initialize Python (Py_Initialize)
* Functions to check if Python is being shutting down (Py_IsFinalizing)
* Functions to shutdown Python (Py_Finalize)

File: Doc/c-api/init.rst (hunk 1 above)
Comment (vstinner) [reply Doc/c-api/init.rst]: Ok, done

Comment (vstinner) [reply Doc/c-api/init.rst]: I updated the doc to specify that we are talking about the **main** interpreter.

File: Doc/c-api/init.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/c-api/init.rst:376]: Any order is good if it is deliberate.

File: Doc/whatsnew/3.13.rst [hunk 3]
```diff
@@ -826,6 +826,10 @@ New Features
   not needed.
   (Contributed by Victor Stinner in :gh:`106004`.)
 
+* Add :c:func:`Py_IsFinalizing` function: check if the Python interpreter is
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.13.rst]:
```suggestion
* Add :c:func:`Py_IsFinalizing` function: check if the main Python interpreter is
```

File: Misc/NEWS.d/next/C API/2023-08-16-17-16-19.gh-issue-108014.wXN3CF.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Add :c:func:`Py_IsFinalizing` function: check if the Python interpreter is
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/C API/2023-08-16-17-16-19.gh-issue-108014.wXN3CF.rst]:
```suggestion
Add :c:func:`Py_IsFinalizing` function: check if the main Python interpreter is
```

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Thanks for the review @ericsnowcurrently and @serhiy-storchaka!

**Comment by vstinner** [comment]
Follow-up: @wjakob asks to add this new function to the limited C API version 3.13.

**Pull request merged**